src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,256 @@
+      int64_t write_bw = info->max_write_bytes * time_in_ms;
+      int64_t read_ops = info->max_read_ops * time_in_ms;
+      int64_t read_bw = info->max_read_bytes * time_in_ms;
+      it.read.ops = std::min(info->max_read_ops * fixed_float, read_ops + it.read.ops);
+      it.read.bw = std::min(info->max_read_bytes * fixed_float, read_bw + it.read.bw);
+      it.write.ops = std::min(info->max_write_ops * fixed_float, write_ops + it.write.ops);
+      it.write.bw = std::min(info->max_write_bytes * fixed_float, write_bw + it.write.bw);
+    }
+  }
+
+  bool should_rate_limit_write(qos_entry& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((it.write_ops() - 1 < 0) && (ops_limit > 0)) ||
+      (it.write_bytes() < 0 && bw_limit > 0))
+    {
+      return true;
+    }
+
+    // we don't want to reduce ops' tokens if we've rejected it.
+    it.write.ops -= fixed_float_rgw_ratelimit;
+    return false;
+  }
+
+  public:
+    RateLimiter(const RateLimiter&) = delete;
+    RateLimiter& operator =(const RateLimiter&) = delete;
+    RateLimiter(RateLimiter&&) = delete;
+    RateLimiter& operator =(RateLimiter&&) = delete;
+    RateLimiter() = delete;
+    RateLimiter(std::atomic_bool *replacing, std::condition_variable *cv)
+      : replacing(replacing), cv(cv)
+    {
+      // prevents rehash, so no iterators invalidation
+      qos_entries.max_load_factor(1000);
+    };
+
+    bool should_rate_limit(const char *method, const std::string& key, ceph::coarse_real_time curr_timestamp, const RGWQoSInfo* qos_info) {
+      if(key.empty() || key.length() == 1 || !qos_info->enabled)
+        return false;
+      bool is_read = is_read_op(method);
+      // it is possible that some requests will have the same ts, because we are using coarse real time
+      auto& it = find_or_create(key);
+      auto curr_ts = curr_timestamp.time_since_epoch();
+      std::unique_lock lock(it.ts_lock);
+      increase_tokens(it, curr_ts, qos_info);
+      if(is_read)
+        return should_rate_limit_read(it, qos_info->max_read_ops, qos_info->max_read_bytes);
+      return should_rate_limit_write(it, qos_info->max_write_ops, qos_info->max_write_bytes);
+    }
+    void giveback_tokens(const char *method, const std::string& key)
+    {
+      bool is_read = is_read_op(method);
+      auto& it = find_or_create(key);
+      std::unique_lock lock(it.ts_lock);
+      if(is_read) {
+        it.read.ops += fixed_float_rgw_ratelimit;
+      } else {
+        it.write.ops += fixed_float_rgw_ratelimit;
+      }
+      
+    }
+    void increase_bw(const char *method,const std::string& key, int64_t amount, const RGWQoSInfo* info, const DoutPrefixProvider* dpp) {
+      if(key.empty() || key.length() == 1 || !info->enabled)
+        return;
+      bool is_read = is_read_op(method);
+      if ((is_read && !info->max_read_bytes) || (!is_read && !info->max_write_bytes))
```

Comment:
```suggestion
      if ((is_read && info->max_read_bytes == 0) || (!is_read && info->max_write_bytes == 0))
```
would be even better if you used a constant instead of zero here (and in all other places)